### PR TITLE
API endpoint for adding offline transactions

### PR DESCRIPTION
**Pull request by knivets**

### CODE REVIEW

File: saas/api/transactions.py [hunk 1]
```diff
@@ -390,6 +392,52 @@ def list(self, request, *args, **kwargs):
...
 
+class ImportTransactionsAPIView(ProviderMixin, APIView):
+    """
+    Insert transactions that were done offline for the purpose of computing
+    accurate metrics.
+
+    """
+
+    def post(self, request, *args, **kwargs):
+        form = ImportTransactionForm(request.data)
```
Comment (smirolo) [inline saas/api/transactions.py]: Is this not a `Serializer` instead?

File: saas/api/transactions.py [hunk 2]
```diff
@@ -390,6 +392,52 @@ def list(self, request, *args, **kwargs):
...
+        if form.is_valid():
```
Comment (smirolo) [inline saas/api/transactions.py]: Take a look at [rest_framework/mixins.py](https://github.com/encode/django-rest-framework/blob/90ed2c1ef7c652ce0a98075c422d25a632d62507/rest_framework/mixins.py#L18) to see how code should be written. Override `perform_create` to fit with the DRF library patterns.

**Comment by knivets** [comment]
Right, I pushed the commit from #123 by mistake. The commit is functional. Should we develop those 2 features as one PR or should I revert this?

**Comment by smirolo** [comment]
Push both as one PR. It shouldn't take too long to complete so let's not bother with reverting.

File: saas/static/js/djaodjin-saas-angular.js [hunk 3]
```diff
@@ -35,7 +35,23 @@ angular.module("saasFilters", [])
...
+                    currency: currency,
+                    useGrouping: false,
+                    minimumFractionDigits: 0,
+                    maximumFractionDigits: 0
+                }).format(1).replace('1', '')
+            }
+            return '';
+        }
+    })
+    .filter("humanizeCell", function(currencyFilter, numberFilter, currencyToSymbolFilter) {
```
[the remark this answers is not in the document]
Comment (knivets) [reply saas/static/js/djaodjin-saas-angular.js]: A separate filter is needed in a places where just a unit is displayed, like in djaoapp.

File: saas/templates/saas/billing/transactions.html [hunk 4]
```diff
@@ -4,7 +4,7 @@
 <div ng-app="saasApp">
   <div ng-controller="transactionListCtrl">
     <section id="total">
-      <strong>Balance</strong> <span>[[items.balance | humanizeCell : '$' : 0.01]]</span>
+      <strong>Balance</strong> <span>[[items.balance | humanizeCell : 'usd' : 0.01]]</span>
```
[the remark this answers is not in the document]
Comment (knivets) [reply saas/templates/saas/billing/transactions.html]:
```
.filter("humanizeCell", function(currencyFilter, numberFilter, currencyToSymbolFilter) {
        "use strict";

        return function(cell, unit, scale, abbreviate) {
            if(typeof abbreviate === "undefined"){
                abbreviate = true;
            }    
            scale = scale || 1;
            var value = cell * scale;
            if(unit) {
                var symbol = currencyToSymbolFilter(unit)
                return currencyFilter(value, symbol, 2);
            }    
            return numberFilter(value);
        };   

    })
```
there is a default value for the 3rd parameter.

File: saas/api/metrics.py [hunk 5]
```diff
@@ -240,32 +250,40 @@ def get(self, request, *args, **kwargs):
...
+        _, refund_amounts, refund_unit = aggregate_transactions_by_period(
             self.provider, Transaction.REFUND,
             orig='dest', dest='dest',
             date_periods=dates)
 
+        units = get_different_units(table_unit, payments_unit, refund_unit)
+        if units:
+            LOGGER.error("different units: %s", units)
+
+        if table_unit:
```
[the remark this answers is not in the document]
Comment (knivets) [reply saas/api/metrics.py]: if `table_unit` is None then it doesn't make sense to use it in comparison. We are only interested in non-`None` units and the ones that differ. That's why I create a set (to remove identical values) first and then remove `None` values: if we're left with `len(units) > 1` that means that we have different unit values.

File: saas/managers/metrics.py [hunk 6]
```diff
@@ -344,21 +344,26 @@ def active_subscribers(plan, from_date=None, tz=None):
 def abs_monthly_balances(organization=None, account=None, like_account=None,
                          until=None, step_months=1, tz=None):
     #pylint:disable=invalid-name,too-many-arguments
-    return [(item[0], abs(item[1]), item[2]) for item in monthly_balances(
-        organization=organization, account=account, like_account=like_account,
-        until=until, step_months=step_months, tz=tz)]
+    balances, unit = monthly_balances(organization=organization,
+        account=account, like_account=like_account,
+        until=until, step_months=step_months, tz=tz)
+    return [(item[0], abs(item[1])) for item in balances]
```
Comment (smirolo) [inline saas/managers/metrics.py]: Missing return of the `unit`?

File: saas/templates/saas/billing/transactions.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (smirolo) [reply saas/templates/saas/billing/transactions.html]: OK. So currencyToSymbolFilter is not really a filter but a boolean (`abbreviate`?). It does seem like `abbreviate` is used... I will have to see the code in action to understand what is happening.

[the remark this answers is not in the document]
Comment (knivets) [reply saas/templates/saas/billing/transactions.html]: I think you're mixing a dependency injection parameters (`currencyFilter, numberFilter, currencyToSymbolFilter`) with actual filter parameters (`cell, unit, scale, abbreviate`)

**Pull request merged**